File: src/libsokoengine/game/sokoban_tessellation.cpp
#include "sokoban_tessellation.hpp"

#include "pusher_step.hpp"
#include "snapshot.hpp"

#include <map>
#include <memory>

using std::invalid_argument;
using std::string;

namespace sokoengine {
namespace game {

using implementation::direction_str;

const Directions &SokobanTessellation::legal_directions() const {
  static const Directions retv = {
    Direction::LEFT, Direction::RIGHT, Direction::UP, Direction::DOWN};
  return retv;
}

position_t SokobanTessellation::neighbor_position(
  position_t       position,
  const Direction &direction,
  board_size_t     width,
  board_size_t     height
) const {
  position_t row = index_y(position, width), column = index_x(position, width);
  switch (direction) {
    case Direction::LEFT:
      column += -1;
      break;
    case Direction::UP:
      row += -1;
      break;
    case Direction::RIGHT:
      column += 1;
      break;
    case Direction::DOWN:
      row += 1;
      break;
    default:
      throw invalid_argument(
        "Unsupported direction " + direction_str(direction)
        + " for SokobanTessellation!"
      );
  }

  if (is_on_board_2d(column, row, width, height))
    return index_1d(column, row, width);
  else
    return Config::NO_POS;
}

char SokobanTessellation::pusher_step_to_char(const PusherStep &rv) const {
  switch (rv.direction()) {
    case Direction::LEFT:
      return rv.is_push_or_pull() ? io::Snapshot::L : io::Snapshot::l;
    case Direction::UP:
      return rv.is_push_or_pull() ? io::Snapshot::U : io::Snapshot::u;
    case Direction::RIGHT:
      return rv.is_push_or_pull() ? io::Snapshot::R : io::Snapshot::r;
    case Direction::DOWN:
      return rv.is_push_or_pull() ? io::Snapshot::D : io::Snapshot::d;
    default:
      throw invalid_argument(
        "Illegal PusherStep direction " + direction_str(rv.direction())
        + " in SokobanTessellation!"

      );
  }
}

PusherStep SokobanTessellation::char_to_pusher_step(char rv) const {
  switch (rv) {
    case io::Snapshot::l:
      return PusherStep(Direction::LEFT, Config::NO_ID);
    case io::Snapshot::L:
      return PusherStep(Direction::LEFT, Config::DEFAULT_ID);
    case io::Snapshot::u:
      return PusherStep(Direction::UP, Config::NO_ID);
    case io::Snapshot::U:
      return PusherStep(Direction::UP, Config::DEFAULT_ID);
    case io::Snapshot::r:
      return PusherStep(Direction::RIGHT, Config::NO_ID);
    case io::Snapshot::R:
      return PusherStep(Direction::RIGHT, Config::DEFAULT_ID);
    case io::Snapshot::d:
      return PusherStep(Direction::DOWN, Config::NO_ID);
    case io::Snapshot::D:
      return PusherStep(Direction::DOWN, Config::DEFAULT_ID);
    default:
      throw invalid_argument(
        string("Illegal PusherStep character '") + rv + "' in SokobanTessellation!"
      );
  }
}

} // namespace game
} // namespace sokoengine

File: src/sokoenginepy/game/sokoban_tessellation.py
from __future__ import annotations

from typing import Dict, Final, List, Optional, Tuple

from ..io import Snapshot
from .base_tessellation import BaseTessellation
from .config import Config, Direction
from .coordinate_helpers import index_1d, index_column, index_row, is_on_board_2d
from .utilities import inverted


class SokobanTessellation(BaseTessellation):
    """
    Tessellation for Sokoban game variant.

    Board is laid out on squares.

    Direction <-> character mapping:

    ====  =====  ====  ====
    LEFT  RIGHT  UP    DOWN
    ====  =====  ====  ====
    l, L  r, R   u, U  d, D
    ====  =====  ====  ====
    """

    _LEGAL_DIRECTIONS: Final[Tuple[Direction, ...]] = (
        Direction.LEFT,
        Direction.RIGHT,
        Direction.UP,
        Direction.DOWN,
    )

    _CHR_TO_PUSHER_STEP: Final[Dict[str, Tuple[Direction, bool]]] = {
        Snapshot.l: (Direction.LEFT, False),
        Snapshot.L: (Direction.LEFT, True),
        Snapshot.r: (Direction.RIGHT, False),
        Snapshot.R: (Direction.RIGHT, True),
        Snapshot.u: (Direction.UP, False),
        Snapshot.U: (Direction.UP, True),
        Snapshot.d: (Direction.DOWN, False),
        Snapshot.D: (Direction.DOWN, True),
    }

    _PUSHER_STEP_TO_CHR: Final[Dict[Tuple[Direction, bool], str]] = inverted(
        _CHR_TO_PUSHER_STEP
    )

    _NEIGHBOR_SHIFT: Final[List[Tuple[Optional[int], Optional[int]]]] = [
        (-1, 0),  # UP
        (None, None),
        (0, 1),  # RIGHT
        (None, None),
        (1, 0),  # DOWN
        (None, None),
        (0, -1),  # LEFT
        (None, None),
    ]

    def neighbor_position(
        self, position: int, direction: Direction, board_width: int, board_height: int
    ) -> int:
        if position < 0:
            raise IndexError(f"Position {position} is invalid value!")

        if board_width < 0:
            raise ValueError(f"Board width {board_width} is invalid value!")

        if board_height < 0:
            raise ValueError(f"Board height {board_height} is invalid value!")

        row = index_row(position, board_width)
        column = index_column(position, board_width)
        row_shift, column_shift = self._NEIGHBOR_SHIFT[direction.value]

        if row_shift is None or column_shift is None:
            raise ValueError(direction)

        row += row_shift
        column += column_shift

        if is_on_board_2d(column, row, board_width, board_height):
            return index_1d(column, row, board_width)

        return Config.NO_POS
